addon/components/base/bs-form.js

```diff
@@ -300,27 +300,23 @@ export default Component.extend({
     this.get('onBefore')(model);
 
     RSVP.resolve(this.get('hasValidator') ? this.validate(this.get('model')) : null)
-      .then((r) => {
-        RSVP.resolve(this.get('onSubmit')(model, r))
-          .finally(() => {
-            if (!this.get('isDestroyed')) {
-              if (this.get('pendingSubmissions') === 1) {
-                this.set('isSubmitting', false);
-              }
-              this.decrementProperty('pendingSubmissions');
-            }
-          });
-      })
-      .catch((err) => {
-        if (!this.get('isDestroyed')) {
+      .then(
+        (r) => {
+          return this.get('onSubmit')(model, r);
+        },
+        (err) => {
           this.set('showAllValidations', true);
 
+          return this.get('onInvalid')(model, err);
```

Comment:
Setting `showAllValidations` to `true` before `onInvalid` fulfills is only for not introducing a breaking change. IMO it would be better to now show validation errors before onInvalid resolves. Maybe value onInvalid resolves with should control if validation errors will be should be shown at all? Let me know if you would except such a breaking change.

Reply:
> IMO it would be better to now show validation errors before onInvalid resolves. Maybe value onInvalid resolves with should control if validation errors will be should be shown at all?

Makes sense!

> Let me know if you would except such a breaking change.

Yeah, not sure about it. I *guess* it should not make any problems in practice, as currently the result of `onInvalid` is not able to control the validations, so showing the validations after the promise has resolved would only cause a slight timing change I guess, nothing which really breaks things from the user's perspective. It could maybe cause some failing tests though!?

An alternative would be to create a tracking issue for this, and implement it for the next major version bump. Which might be happening rather soon maybe, if e-b drops support for Ember CLI < 2.12 (to be able to upgrade to Babel 7). 

*Tbh, the semver "contract" regarding Ember CLI is quite fuzzy (a problem with all addons), as it nowhere states which versions it supports. But as e-b support Ember 2.3, I would think it would be reasonable to expect the same for CLI!? Hm, getting a bit off-topic...* 😝

Thoughts?

Reply:
> It could maybe cause some failing tests though!?

I think this would be only the case in very few rare cases, cause mostly cases await all promises to be settled after triggering a user interaction (e.g. `await click('button[type="submit"]');`). In that cases `onInvalid` will be resolved anyways before assertion is done. This change might only fail tests if they await an interim state using `waitUntil` or `waitFor`. In my experience that's not used quite often yet.

> An alternative would be to create a tracking issue for this, and implement it for the next major version bump. 
It's your call. Could create another pull request with necessary changes after this one has landed and you are free to merge it at any time.

> Tbh, the semver "contract" regarding Ember CLI is quite fuzzy (a problem with all addons), as it nowhere states which versions it supports. But as e-b support Ember 2.3, I would think it would be reasonable to expect the same for CLI!? Hm, getting a bit off-topic... stuck_out_tongue_closed_eyes

IMO an ember version should be considered as a specific bundle of `ember-source`, `ember-cli` and `ember-data`. Furthermore I would argue that supporting a specific version should mean that it works fine with default blueprint for that ember version. You could always extend that general statement by saying that it also supports older versions of `ember-source`.